### PR TITLE
[core] Fix wrong local resource view in raylet

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
When gcs broad cast node resource change, raylet will use that to update local node as well which will lead to local node instance and nodes_ inconsistent.

1. local node has used all some pg resource
2. gcs broadcast node resources
3. local node now have resources
4. scheduler picks local node
5. local node can't schedule the task
6. since there is only one type of job and local nodes hasn't finished any tasks so it'll go to step 4 ==> hangs

## Related issue number
#19438 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
I don't have a good idea about how to test this since it's related to grpc and raylet.
I also don't know how to construct a unit test for this, since it's has a lot of components involved. :( 
Any good suggestions about how to test this one?

If the code looks good, I'd prefer to merge it first so that we can observe it from the nightly test to see whether it's working or not.

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [inline src/ray/raylet/node_manager.cc:799]: Only this line.

**Comment by iycheng** [comment]
I'm also building wheel to test it in testing env to make sure it's working.

**Comment by iycheng** [comment]
still hangs sometimes.. investigating

**Comment by iycheng** [comment]
I've run it several times in test env and things look good.

**Comment by iycheng** [comment]
This PR might be related to this task https://github.com/ray-project/ray/issues/19326 @scv119

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:1483]: Can you comment what 0 means here?  (in the code)

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:912]: So, when the local resources are updated, `ScheduleAndDispatchTasks ` is invoked? Can you point me where that happens?

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:1499]: all changes underneath is just refactoring right?

**Review by rkooo567 — APPROVED** [review]
Nice catch! So, the bug is the local information becomes inconsistent because we update the local resources by the resource update message from GCS, which is more stale right? 

I think how we can test this is to ensure local resources are never updated by GCS reports from cpp unit tests, but it seems not simple with the current code structure (maybe we should move all resource update code path into other class or something).

**Comment by rkooo567** [comment]
Btw, the CI seems to have the compiler failure. Can you fix that? (I'd like to see if it will pass all CI tests)

**Comment by iycheng** [comment]
Thanks, @rkooo567 
I actually implemented a simple asio defer execution framework for testing... but then realize that I can reproduce it more simply by just increasing reporting interval

I'll update this PR with py test and then maybe have another PR for the testing framework which might be useful for other usage.

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:912]: Local resources are updated by local code. https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L1634

Right now, only pg will change local resource

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:1499]: just make it looks cleaner.

**Comment by iycheng** [comment]
> Nice catch! So, the bug is the local information becomes inconsistent because we update the local resources by the resource update message from GCS, which is more stale right?

Actually, it's an existing bug but easy to trigger by grpc resource broadcasting. 


Difference between grpc based and redis based: 
redis based: always send it when node resource changed.
grpc based: send it with resource broadcasting (100ms)


For redis based one, since we haven't eat all the local resources, it's ok since the next schedule will update it to the correct one.

For grpc based one, since it'll use 100ms, all local resource will be eat and the next update will change the local resource to be the one with full resources. So it'll hang since ray will always select the local instead of remote.

File: src/ray/raylet/node_manager.cc
Comment (mwtian) [inline src/ray/raylet/node_manager.cc:1485]: Handle `resource_usage_batch.seq_no() > next_resource_seq_no_` and `resource_usage_batch.seq_no() < next_resource_seq_no_` separately?

File: src/ray/raylet/node_manager.cc
Comment (mwtian) [inline src/ray/raylet/node_manager.cc:1502]: nit: maybe move the comment to be above the if block.

**Review by mwtian — APPROVED** [review]
Good catch!

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:1485]: Sounds good. I previously wanted to handle this one after we have lightweight resource broadcasting. But I guess here probably just discard the stale seq_no

**Comment by iycheng** [comment]
It looks like my test might polluted the test since this build actually passed https://buildkite.com/ray-project/ray-builders-pr/builds/17447

**Comment by iycheng** [comment]
test failure looks related to the master failure. holding this pr

**Comment by iycheng** [comment]
verified the failure looks the same as trunk failure and also windows failure (the one before merge) is the same as trunk failure.
lint failure is also not related to this pr and failed on the trunk as well.

I need this to run on nightly test for several days.

**Pull request merged**
